modules/tensor_mechanics/src/materials/ADMultiplePowerLawCreepStressUpdate.C

```diff
@@ -0,0 +1,141 @@
+    paramError("stress_thresholds",
+               "The number of creep models must be the number of stress thresholds minute. That "
+               "is, one creep model will be valid between two stress thresholds");
+
+  for (unsigned int i = 0; i < _number_of_models; i++)
+    if (_start_time < this->_app.getStartTime() && (std::trunc(_m_exponent[i]) != _m_exponent[i]))
+      paramError("start_time",
+                 "Start time must be equal to or greater than the Executioner start_time if a "
+                 "non-integer m_exponent is used");
+}
+
+void
+ADMultiplePowerLawCreepStressUpdate::computeStressInitialize(
+    const ADReal & effective_trial_stress, const ADRankFourTensor & /*elasticity_tensor*/)
+{
+  _stress_index = stressIndex(effective_trial_stress);
+
+  if (_temperature)
+    _exponential =
+        std::exp(-_activation_energy[_stress_index] / (_gas_constant * (*_temperature)[_qp]));
+
+  _exp_time = std::pow(_t - _start_time, _m_exponent[_stress_index]);
+}
+
+ADReal
+ADMultiplePowerLawCreepStressUpdate::computeResidual(const ADReal & effective_trial_stress,
+                                                     const ADReal & scalar)
+{
+  const ADReal stress_delta = effective_trial_stress - _three_shear_modulus * scalar;
+  const ADReal creep_rate = _coefficient[_stress_index] *
+                            std::pow(stress_delta, _n_exponent[_stress_index]) * _exponential *
+                            _exp_time;
+  return creep_rate * _dt - scalar;
+}
+
+ADReal
+ADMultiplePowerLawCreepStressUpdate::computeDerivative(const ADReal & effective_trial_stress,
+                                                       const ADReal & scalar)
+{
+  const ADReal stress_delta = effective_trial_stress - _three_shear_modulus * scalar;
+  const ADReal creep_rate_derivative =
+      -_coefficient[_stress_index] * _three_shear_modulus * _n_exponent[_stress_index] *
+      std::pow(stress_delta, _n_exponent[_stress_index] - 1.0) * _exponential * _exp_time;
+  return creep_rate_derivative * _dt - 1.0;
+}
+
+Real
+ADMultiplePowerLawCreepStressUpdate::computeStrainEnergyRateDensity(
+    const ADMaterialProperty<RankTwoTensor> & stress,
+    const ADMaterialProperty<RankTwoTensor> & strain_rate)
+{
+  if (_n_exponent[_stress_index] <= 1)
+    return 0.0;
+
+  Real creep_factor = _n_exponent[_stress_index] / (_n_exponent[_stress_index] + 1);
+
+  return MetaPhysicL::raw_value(creep_factor * stress[_qp].doubleContraction((strain_rate)[_qp]));
+}
+
+bool
+ADMultiplePowerLawCreepStressUpdate::substeppingCapabilityEnabled()
+{
+  return getParam<bool>("use_substep");
+}
+
+std::size_t
+ADMultiplePowerLawCreepStressUpdate::stressIndex(const ADReal & effective_trial_stress)
+{
+  for (unsigned int i = 0; i < _number_of_models; i++)
+    if (effective_trial_stress >= _stress_thresholds[i] &&
+        effective_trial_stress < _stress_thresholds[i + 1])
+      return i;
+
+  return _number_of_models - 1;
```

Comment:
Should you use the actual stress (returned stress) instead of trial stress to determine which model to use?

Reply:
The code does assume that the trial stress and the actual stress are close relative to the stress thresholds provided by the user -- which, for the case of creep, should be true for a simulation with acceptable creep integration error. Choosing the index prior to the return mapping loop makes the all the quantities consistent throughout the return mapping iterations. You could think of cases where, if we update the index with the returned stress and it crosses stress thresholds during the return mapping iterations (potentially back and forth?), it could give rise to convergence issues.